sapio-psbt/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright Judica, Inc 2022
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+//  License, v. 2.0. If a copy of the MPL was not distributed with this
+//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use bitcoin::consensus::serialize;
+use bitcoin::schnorr::TapTweak;
+use bitcoin::secp256k1::rand::Rng;
+use bitcoin::secp256k1::{rand, All};
+use bitcoin::util::bip32::{ExtendedPubKey, KeySource};
+use bitcoin::util::sighash::Prevouts;
+use bitcoin::util::taproot::TapLeafHash;
+use bitcoin::util::taproot::TapSighashHash;
+use bitcoin::XOnlyPublicKey;
+use bitcoin::{
+    psbt::PartiallySignedTransaction, secp256k1::Secp256k1, util::bip32::ExtendedPrivKey,
+};
+use bitcoin::{KeyPair, TxOut};
+use bitcoin::{Network, SchnorrSig};
+use miniscript::psbt::PsbtExt;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::error::Error;
+use std::fmt::Display;
+use std::str::FromStr;
+
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum PSBTApi {
+    Finished {
+        completed: bool,
+        hex: String,
+    },
+    NotFinished {
+        completed: bool,
+        psbt: String,
+        error: String,
+        errors: Vec<String>,
```

Comment:
o.O

Reply:
hah, fair.

We could preserve the structure of the errors from https://docs.rs/miniscript/7.0.0/miniscript/psbt/enum.Error.html, but these don't implement Serialize (maybe they should?) and also aren't a standard error type understandable by other software. So they're more provided for convenience rather than something a consumer can use -- this particular function is mostly to be used by people wanting to output JSONs

Reply:
(this is also why we do untagged repr, so that it matches the APIs from e.g. core)